packages/hardhat/test/e2e/crab-Migration.ts

```diff
@@ -273,4 +322,702 @@ describe("Crab Migration", function () {
+            await initialize(d5.address)
+            await decreaseCR1(ethers.utils.parseEther('10'))
+            await expect(crabMigration.connect(d5).flashMigrateFromV1toV2(one, 0)).to.be.revertedWith("M11");
+            await increaseCR1(ethers.utils.parseEther('10'))
+        })
+
+        it("Should fail if _ethToBorrow or _withdrawMaxEthToPay is passed as 0", async () => {
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, one, 0, 0)).to.be.revertedWith("M8");
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, one, one, 0)).to.be.revertedWith("M8");
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, one, 0, one)).to.be.revertedWith("M8");
+        })
+
+        it("Should migrate d4 when CR1 > CR2", async () => {
+            let gasPaid = BigNumber.from(0)
+            await initialize(d4.address)
+            await increaseCR1(ethers.utils.parseEther('10'))
+            const [isFlashMigrate, ethNeededForV2, , ethToGetFromV1] = await crabMigration.flashMigrationDetails(crabV1SharesBefore)
+            const crabV2SharesBefore = await crabStrategyV2.balanceOf(d4.address)
+
+            expect(wdiv(collatV1, shortV1).gt(wdiv(collatV2, shortV2))).to.be.true
+            expect(isFlashMigrate).to.be.true
+            expect(ethToGetFromV1.gt(ethNeededForV2)).to.be.true
+
+            const excessEth = ethToGetFromV1.sub(ethNeededForV2)
+            const depositShare = wdiv(ethNeededForV2, collatV2.add(ethNeededForV2))
+            const expectedV2Shares = wdiv(wmul(depositShare, crabV2Supply), one.sub(depositShare))
+
+            const tx1 = await crabStrategyV1.connect(d4).approve(crabMigration.address, crabV1SharesBefore);
+            gasPaid = await getGasPaid(tx1)
+
+            const tx2 = await crabMigration.connect(d4).flashMigrateFromV1toV2(crabV1SharesBefore, 0);
+            gasPaid = gasPaid.add(await getGasPaid(tx2))
+
+            const crabV1SharesAfter = await crabStrategyV1.balanceOf(d4.address)
+            const crabV2SharesAfter = await crabStrategyV2.balanceOf(d4.address)
+            const crabV1SharesInMigration = await crabStrategyV1.balanceOf(crabMigration.address)
+            const crabV2SharesInMigration = await crabStrategyV2.balanceOf(crabMigration.address)
+            const userEthBalanceAfter = await provider.getBalance(d4.address)
+            const squeethBalance = await oSqth.balanceOf(crabMigration.address)
+
+            expect(crabV1SharesAfter).to.be.equal('0')
+            expect(crabV1SharesInMigration).to.be.equal('0')
+            expect(crabV2SharesBefore).to.be.equal('0')
+            expect(crabV2SharesAfter).to.be.equal(expectedV2Shares)
+            expect(userEthBalanceAfter).to.be.equal(userEthBalance.add(excessEth).sub(gasPaid))
+            expect(crabV2SharesInMigration).to.be.equal('0')
+            expect(squeethBalance).to.be.equal('0')
```

Comment:
Does the user ever have a change in squeeth balance (ie are they ever sent 1 wei?)